File: BOJ(Java 공부용)/1963.java
import java.io.*;
import java.util.*;
import static java.lang.Math.sqrt;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int[] notprime = new int[10001];
        for(int i=2;i<=sqrt(10000);i++) {
            if(notprime[i] == 1) continue;
            for(int j=i*i;j<=10000;j+=i) {
                notprime[j] = 1;
            }
        }
        ArrayList<Integer> v = new ArrayList<>();
        for(int i=2;i<10000;i++)  {
            if(notprime[i] ==0 ) v.add(i);
        }
        int t= Integer.parseInt(st.nextToken());
        for(int tc = 0 ;tc<t;tc++) {
            st = new StringTokenizer(br.readLine());
            int n= Integer.parseInt(st.nextToken()), m = Integer.parseInt(st.nextToken());
            int ans = bfs(n,m,v);
            bw.write(ans==-1?"Impossible":ans+"\n");
        }
        bw.flush();
        bw.close();
    }
    private static int bfs(int st,int en,ArrayList<Integer> v) {
        boolean[] visited =new boolean[10001];
        visited[st] = true;
        ArrayDeque<p> q = new ArrayDeque<>();
        q.add(new p(st,0));
        int temp =-1;
        while(!q.isEmpty()) {
            int num = q.peek().num;
            int cnt = q.peek().cnt;
            q.poll();
            if(num == en) {
                temp = cnt;
                break;
            }
            StringBuilder sb = new StringBuilder();
            sb.append(num);
            for(int i=0;i<sb.length();i++) {
                char c = sb.charAt(i);
                for(int j=0;j<10;j++) {
                    if(c == j+'0') continue;
                    StringBuilder ssb = new StringBuilder();
                    ssb.append(num);
                    ssb.setCharAt(i, (char) (j+'0'));
                    int newnum = Integer.parseInt(ssb.toString());
                    int idx = lowerbound(0,v.size(),newnum,v);
                    if(idx == v.size()) continue;
                    if(v.get(idx) == newnum && !visited[newnum]  && newnum>=1000 && newnum<10000) {
                        q.add(new p(newnum,cnt+1));
                        visited[newnum] = true;
                    }
                }
            }
        }
        return temp;
    }
    private static class p {
        int num,cnt;
        private p(int num,int cnt) {
            this.num =num;
            this.cnt = cnt;
        }
    }
    private static int lowerbound(int left, int right, int num, ArrayList<Integer> v) {
        while(left <right) {
            int mid = (left+right) /2 ;
            if(v.get(mid) < num) left = mid+1;
            else right = mid;
        }
        return right;
    }
}

File: BOJ(Python 공부용)/1963.py
import sys
from collections import deque
from math import sqrt

input = sys.stdin
sys.setrecursionlimit(10000000)


def lowerbound(left, right, num):
    while left < right:
        mid = (left + right) // 2
        if v[mid] < num:
            left = mid + 1
        else:
            right = mid
    return right


def makenum(str):
    sum = 0
    idx = 0
    for i in range(len(str) - 1, -1, -1):
        sum += int(str[i]) * int(pow(10, idx))
        idx += 1
    return sum


def bfs(st, en):
    visited = [False] * 10001
    visited[st] = True
    q = deque()
    q.append((st, 0))
    temp = -1
    while q:
        num, cnt = q.popleft()
        if num == en:
            temp = cnt
            break
        sb = list(str(num))
        for i in range(len(sb)):
            c = sb[i]
            for j in range(10):
                if int(c) == j: continue
                ssb = list(str(num))
                ssb[i] = str(j)
                newnum = makenum(ssb)
                idx = lowerbound(0, len(v), newnum)
                if idx == len(v): continue
                if v[idx] == newnum and not visited[newnum] and 1000 <= newnum < 10000:
                    q.append((newnum, cnt + 1))
                    visited[newnum] = True

    return temp


notprime = [False] * 10001
for i in range(2, int(sqrt(10000))):
    for j in range(i * i, 10000, i):
        notprime[j] = True

v = []
for i in range(2, 10000):
    if not notprime[i]:
        v.append(i)

t = int(input.readline())
for _ in range(t):
    n, m = map(int, input.readline().split())
    ans = bfs(n, m)
    if ans == -1:
        print("Impossible")
    else:
        print(ans)
